src/net_processing.cpp

```diff
@@ -835,11 +835,13 @@ void PeerManager::FinalizeNode(const CNode& node, bool& fUpdateConnectionTime) {
     LogPrint(BCLog::NET, "Cleared nodestate for peer=%d\n", nodeid);
 }
 
-PeerRef PeerManager::GetPeerRef(NodeId id)
+PeerRef PeerManager::GetPeerRef(NodeId id) const
 {
+    PeerRef ret;
     LOCK(m_peer_mutex);
     auto it = m_peer_map.find(id);
-    return it != m_peer_map.end() ? it->second : nullptr;
+    if (it != m_peer_map.end()) ret = it->second;
+    return ret;
```

Comment:
6bfcef3797d4d105180e325033929684f84cfc08

Why are these changes required? What are benefits to return an empty shared pointer instead `nullptr`-constructed one? Btw, in all caller places the return value is checked for `nullptr`, not being empty.

Reply:
This makes is clear that NRVO can be used, since there's only one return statement. I'm not sure if the previous version would have used copy elision.

Reply:
+1 for considering this!

I'd be curious to know if compilers are allowed to rearrange the code before deciding how many return statements there are. I could see this going either way.

Reply:
Considering the before-change code, to be precise, in C++17 in a return statement, when the operand is a prvalue (the result of the ternary conditional expression), elision of copy/move operations is [mandatory](https://en.cppreference.com/w/cpp/language/copy_elision).

Reply:
Thanks @hebasto! Yes, I think you're right. If I'm reading it correctly, the result of the conditional operator is a prvalue (item 5 in https://en.cppreference.com/w/cpp/language/operator_other#Conditional_operator since 4 doesn't apply), and copy elision is mandatory for returning a prvalue.

I've reverted this.

Reply:
> ... and copy elision is mandatory for returning a prvalue.

Good to have C++17 :)